Qdisk optimization 1

Barkodcz opened the pull request:
Optimization:
1. using 1 write instead of 2
2. move GString to struct from function

Commit:
qdisk: using 1 write instead of 2

* concatenate message length and message, then write

This results in obvious performance improvement, makes message-writes a bit more atomic, and does not worsen disk-buffer guarantees.

Signed-off-by: Balázs Barkó <Balazs.Barko@oneidentity.com>

Commit:
qdisk: using scratch_buffers to serialize the message

reason: didn't need to alloc && free GString every time when we want to write message to file

Signed-off-by: Balázs Barkó <Balazs.Barko@oneidentity.com>

Commit:
qdisk: using scratch_buffers to deserialize messages

Signed-off-by: Balázs Barkó <Balazs.Barko@oneidentity.com>

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -320,8 +320,10 @@ qdisk_push_tail(QDisk *self, GString *record)
+  g_string_truncate(self->temp, 0);
+  g_string_append_len(self->temp, ((gchar *) &record_length), sizeof(record_length));
+  g_string_append_len(self->temp, (record->str), record->len);
+  if (!pwrite_strict(self->fd, self->temp->str, self->temp->len, self->hdr->write_head))
```
I think using a single pwrite() is a good idea in general, because it's one step closer to atomic writes.

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -77,6 +77,7 @@ struct _QDisk
   gint64 file_size;
   QDiskFileHeader *hdr;
   DiskQueueOptions *options;
+  GString *temp;
```
Adding such a state to QDisk is not a good idea in my opinion.

If we needed a string buffer as a class field, we would have to find a more eloquent name for that, but I think we can do better by using scratch buffers instead. They were made exactly for this use case.

For example:
https://github.com/syslog-ng/syslog-ng/blob/e0104231703d5d68091183f24696e83ba014a722/modules/xml/xml.c#L61-L87
I think it would be important to use the mark and reclaim API, because these buffers will grow big.
(`scratch_buffers_mark()`, `scratch_buffers_reclaim_marked()`). `scratch_buffers_alloc()` gives you a pre-allocated, but empty (truncated) `GString`.

The main advantage in this context would be keeping string buffers local on hot paths.
Disk-buffer is a multi-threaded entity, its methods are called from multiple source threads and a destination thread as well. We currently have top-level mutexes that prevent race conditions inside QDisk, but that can eventually be changed. 

Moving a local string buffer into the `QDisk` struct complicates things further when considering the parallel access of this field.

MrAnno commented on modules/diskq/logqueue-disk.h:
```diff
@@ -35,6 +35,7 @@ struct _LogQueueDisk
 {
   LogQueue super;
   QDisk *qdisk;         /* disk based queue */
+  GString *write_serialized;
```
I'd use scratch buffers here (see my previous comment).

MrAnno commented:
I've found one more local string buffer, which we could replace with a scratch buffer:
- `logqueue-disk-reliable.c: _skip_message()`

Commit:
qdisk: using scratch_buffers when skipping messages

_skip_message() is on hot-path, it is used to skip messages from a reliable disk-buffer file when
the corresponding message can be found in the qreliable memory buffer segment.

Signed-off-by: Balázs Barkó <Balazs.Barko@oneidentity.com>

MrAnno commented on modules/diskq/logqueue-disk.c:
```diff
@@ -222,29 +223,32 @@ _free(LogQueue *s)
+  ScratchBuffersMarker marker;
+  GString *read_serialized;
+  read_serialized = scratch_buffers_alloc_and_mark(&marker);
+  g_string_erase(read_serialized, 0, -1);
```
`g_string_erase()` is not necessary, scratch buffers are truncated by the API.

MrAnno reviewed (commented):
Nice, thank you. 👍🏻 

Nitpicking: merging the declaration of `read_serialized`, `write_serialized`, and `msg_buffer` with its first assignment would help readability.

I think we should be more verbose with the commit messages as these commits will have to be reviewed thoroughly and carefully in the end, to make sure we did not break anything.
My suggestion would be the following (I fixed typos too):

----
qdisk: using 1 write instead of 2
* concatenate message length and message, then write

This results in obvious performance improvement, makes message-writes a bit more atomic, and does not worsen disk-buffer guarantees.

----

qdisk: using scratch_buffers to serialize the message

reason: didn't need to alloc && free GString every time when we want to write message to file

----

qdisk: using scratch_buffers to deserialize messages

----

qdisk: using scratch_buffers when skipping messages

_skip_message() is on hot-path, it is used to skip messages from a reliable disk-buffer file when
the corresponding message can be found in the qreliable memory buffer segment.

MrAnno reviewed (approved):
Thanks!
